### DISCUSSION TITLE
Snippets vs. ViewSets vs. ModelAdmin

### DISCUSSION THREAD

**Discussion by th3hamm0r** [wagtail/wagtail#9115] [Ideas]
We are mainly using wagtail's `ModelAdmin` ([docs](https://docs.wagtail.org/en/stable/reference/contrib/modeladmin/index.html)) for our custom models, because they can be setup fast, but are also very easy to extend (e.g. overriding base views, filtering querysets, adding custom urls/views,...).
We've actually never really used [snippets](https://docs.wagtail.org/en/stable/topics/snippets.html) in any of our projects for our custom models, `ModelAdmin` was much easier to extend (thanks to the class itself as the central configuration) and also simple things like placing it somewhere else in the menu works better with the ModelAdmin. There are also some packages outside, which add helpful functionality, eg. drag&drop re-ordering with [wagtail-orderable](https://github.com/elton2048/wagtail-orderable/).
Also working with wagtail's ModelAdmin feels much more natural, as it has many similarities with django's ModelAdmin.

**I'm coming to the point now:**
In the last major releases there has been a lot of refactoring with `ViewSets`, probably mainly to reduce duplicate code and also many features have been implemented for snippets (unfortunately not for ModelAdmin).
Why not focus on refactoring all model-managing code to be based on the ModelAdmin and add features like bulk actions to it?
I assume this has been discussed in the past, and maybe there are some obvious blockers? 😅

Looking at all the existing admins for Images, Documents, Snippets, Media ([wagtailmedia](https://github.com/torchbox/wagtailmedia/)), but also other admin pages like all the report views, they basically all have the same components which are often duplicated for them:
* CRUD views with permission checks
* search
* filters (collections, tags, ...)
* paging

These features are already available in the ModelAdmin.
Of course, if used to manage images for example, probably the views would need more customizations to get a grid view instead of a table view, also to get a live search, and there are probably more obstacles. But it still seems to me as the better option for the future.
If all those models would be migrated to the ModelAdmin, many projects could benefit from that and its improvements in the long run.

Currently, the ModelAdmin is only exclusively registered for one model (similar to django's admin), so migrating reports, which are often different views on models, or with no single model at all, probably requires some additional handling, so they don't get registered to the `AdminURLFinder`, but even that may already work with all the methods already available in the ModelAdmin classes.

Some features that are missing or which could be improved, and which would really benefit from a migration:
* Filters:
Basic boolean and foreign-key filters work well, date filters for example always require some boilerplate code in the backend and the template to get nice date choosers. Those choosers are already used to filter reports, so this part could benefit from refactoring the reports towards ModelAdmin.
* Bulk actions:
  They have been implemented for images, documents, snippets, users and probably other models (wagtailmedia doesn't have it). They could be implemented in the ModelAdmin similar to django's [actions](https://docs.djangoproject.com/en/stable/ref/contrib/admin/actions/) and probably used very easily by all subclasses, similar to the other features.
* Translations, Revisions,...

**Summary:**
I think the new `ViewSets` are very helpful, and I also think that not all admin views can be implemented with `ModelAdmin` (e.g. dashboards, non-db stuff,...), but actually the majority of the currently implemented views could be migrated (images, documents, media, snippets, sites, locales, reports, users, groups... even the page editor 🤯).

**Comment by lb-** [comment]
Note. There is a similar discussion here (on the supercharged snippets discussion)

https://github.com/wagtail/wagtail/discussions/8609#discussioncomment-3184811


Also - request to add bulk actions to ModelAdmin - https://github.com/wagtail/wagtail/issues/2719

**Reply by th3hamm0r** [reply]
Oh, thanks @lb-, I've missed that part of the discussion...
Reading it, I'm still hoping that `ModelAdmin` remains a core feature, just as an easy equivalent to django's `ModelAdmin` in wagtail :pray: 
I really like https://github.com/wagtail/wagtail/discussions/8609#discussioncomment-3220467 and the idea of having viewsets as a low-level API and ModelAdmin on top
